### PR TITLE
arrays can now be printed

### PR DESCRIPTION
**Pull request by krux02**
fixes https://github.com/nim-lang/Nim/issues/5828
```
echo([1,2,3])
```

The result of ``$`` on array of char is now something that might need some discussion. Because by thefault `$` on `char` creates a string of one unescaped character, the resulting string might be of length 1, but with the `'\0'` as only character. If this string is concatenated into the result, the result is literally a string with a terminating `\0` in the middle.

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -1871,6 +1871,27 @@ proc `$` *[Enum: enum](x: Enum): string {.magic: "EnumToStr", noSideEffect.}
   ## a ``$`` operator for a concrete enumeration is provided, this is
   ## used instead. (In other words: *Overwriting* is possible.)
 
+proc `$`*[T: ref](arg: T): string =
```
Comment (Araq) [inline lib/system.nim]: Remove this please.

File: lib/system.nim [hunk 2]
```diff
@@ -1871,6 +1871,27 @@ proc `$` *[Enum: enum](x: Enum): string {.magic: "EnumToStr", noSideEffect.}
...
+  ## The stringify operator to handle the nil value of all ref types.
+  ## Then it forwards to stringify operator of the underlying (value) type.
+  if arg.isNil:
+    "nil"
+  else:
+    "ref " & $arg[]
+
+proc `$`*[T: ptr](arg: T): string =
```
Comment (Araq) [inline lib/system.nim]: Remove this please.

Comment (krux02) [inline lib/system.nim]: yes

**Comment by Araq** [comment]
This seems to fix things on the wrong level. The implicit conversion from array of char to ``cstring`` should perhaps be removed from the language or weakened further.

**Comment by krux02** [comment]
I think the implicit conversion from array of car to ``cstring`` should be removed from the language. The implicit conversion from ``ptr char`` to ``cstring`` can remain in the language. In Nim arrays are value types, the array is not an alias for the pointer to the first element. When I have an array, I can always get the addr of the first element to pass it to a c library, so I don't see the need to have an implicit conversion of ``array[N,char]`` to ``cstring``.

**Comment by Araq** [comment]
Also related: https://github.com/nim-lang/Nim/issues/5861

**Comment by Araq** [comment]
> In Nim arrays are value types, the array is not an alias for the pointer to the first element. When I have an array, I can always get the addr of the first element to pass it to a c library, so I don't see the need to have an implicit conversion of array[N,char] to cstring.

Agreed.

**Comment by krux02** [comment]
I checked what the problems are in the continuous integration test, but appveyor has this message:
```
Build execution time has reached the maximum allowed time for your plan (60 minutes).
```
But I don't see a specific test that doesn't work.

And on travis-ci when I want to see the job log I get an eternal loading indicator. So don't know what the problem is.

So there is nothing in particular that I can fix to improve this pull request. What should I do?

File: lib/system.nim [hunk 3]
```diff
@@ -1871,6 +1871,21 @@ proc `$` *[Enum: enum](x: Enum): string {.magic: "EnumToStr", noSideEffect.}
   ## a ``$`` operator for a concrete enumeration is provided, this is
   ## used instead. (In other words: *Overwriting* is possible.)
 
+proc newString*[N](data: array[N, char]): string {.noSideEffect.} =
```
Comment (Araq) [inline lib/system.nim]: This is just confusing. The type signature clearly suggests that the array is of a fixed size and not treated as null terminated. Use at least a name like ``newStringZ``.

File: lib/system.nim [hunk 4]
```diff
@@ -2401,42 +2416,44 @@ proc `$`*[T: tuple|object](x: T): string =
...
-  result = b
+proc collectionToString[T](x: T, prefix, separator, suffix: string): string =
+  result = prefix
   var firstElement = true
   for value in items(x):
-    if not firstElement: result.add(", ")
+    if firstElement:
+      firstElement = false
+    else:
+      result.add(separator)
```
Comment (Araq) [inline lib/system.nim]: The separator is always ", " so there is no need to generalize it.

File: lib/system/repr.nim [hunk 5]
```diff
@@ -16,27 +16,27 @@ proc reprInt(x: int64): string {.compilerproc.} = return $x
 proc reprFloat(x: float): string {.compilerproc.} = return $x
 
 proc reprPointer(x: pointer): string {.compilerproc.} =
-  var buf: array[0..59, char]
-  discard c_sprintf(buf, "%p", x)
```
Comment (Araq) [inline lib/system/repr.nim]: But this wastes more storage. Why not pass ``addr(buf[0])`` to ``c_sprintf`` instead?

File: lib/system.nim (hunk 3 above)
Comment (krux02) [inline lib/system.nim]: This procedure is for all cases where an array char is converted into a string. So it does what ``$`` on ``array[..., char]`` used to do. I personally don't like to introduce new names for methods when it's not necessary. But I can easily make this procedure work for both, null terminated and non null terminated arrays. What do you think about that?

Comment (Araq) [inline lib/system.nim]: Well actually, ``$`` works for ``cstring`` and what we need is perhaps ``proc toCstring[Idx](a: array[Idx, char]): cstring = cast[cstring](unsafeAddr(a[0]))``

File: lib/system.nim (hunk 4 above)
Comment (krux02) [inline lib/system.nim]: Technically that is true. the separator is always ``","``  where this function is used. The reason I did add the separator as a parameter is, I can imagine very likely that this procedure might become exported from system at some point. and when it does get exported, I would really like it to have the separator as a parameter. It would be the equivalent of [``mkString`` in Scala](http://www.scala-lang.org/api/2.12.2/scala/collection/Seq.html#mkString(start:String,sep:String,end:String):String) and the additional parameter doesn't really cost anything. I remember to like using ``mkString`` in Scala very much. It was an easy to remember name, easy to use and very flexible to make Strings from collections.

File: lib/system/repr.nim (hunk 5 above)
Comment (krux02) [inline lib/system/repr.nim]: I had problems with one of the methods, that they were executed on the VM, and I got the error in ``newString`` that casting is not allowed on the VM. I fixed ``newString`` to work on the VM, but I didn't like the idea to run a loop over characters on the VM, so I tried to replace the instances where this happened to work on strings directly rather than creating an array first and then convert it.

File: lib/system.nim (hunk 3 above)
Comment (krux02) [inline lib/system.nim]: Well I don't generally have a problem with  ``toCstring``, but I also don't see a direct necessity to do so. It's a new name that needs to be introduced and remembered, and the cast works fine in my opinion.

Comment (krux02) [inline lib/system.nim]: I just found out that ``$buffer.cstring`` does already work, so in fact I could simply remove the new ``newString`` proc.

**Comment by krux02** [comment]
Well I don't get this problem. I don't have a Windows Machine and the error happens in Windows only code:

```
Failure: reExitcodesDiffer
Expected:
exitcode: 0

Gotten:
exitcode: 1

Output:
Traceback (most recent call last)
tioselectors.nim(615)    tioselectors
tioselectors.nim(521)    socket_notification_test
system.nim(3601)         failedAssertImpl
system.nim(3593)         raiseAssert
system.nim(2708)         sysFatal
Error: unhandled exception: 
```

File: lib/system/repr.nim (hunk 5 above)
Comment (krux02) [inline lib/system/repr.nim]: I reverted this back to use an Array.

**Comment by dom96** [comment]
Could we get a test for this as well?

**Comment by krux02** [comment]
The problem is, that I am a bit blocked right now, I don't have a windows machine to test this.

**Comment by dom96** [comment]
Maybe you could set up a Windows VM?

**Comment by krux02** [comment]
Well I have a windows license from University, but in order to get that windows version I need to download it with this "secure" download manager that only works on windows. Well I can see what I can do. But don't expect any updates here the next few days. I am a bit busy (paper submission deadline).

**Comment by jyapayne** [comment]
@krux02 There are free virtual machines offered by Microsoft [here](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines).

**Comment by TiberiumN** [comment]
@jyapayne but you need server for it...

**Comment by krux02** [comment]
I don't get the error from the failed text on AppVeyor

```
tests\system\toString.nim(17, 3) Error: symbol expected, but found: @[23, 45]
```
https://ci.appveyor.com/project/Araq/nim/build/1288#L3013

that column does not exist

https://github.com/krux02/Nim/blob/7ce61f7cd15fe8e80d5221dc2386692a25478aef/tests/system/toString.nim#L17

What should I do about this?

**Comment by krux02** [comment]
I think this pull request is now ready for review.

**Comment by krux02** [comment]
Build execution time has reached the maximum allowed time for your plan (60 minutes).

Well that is not my fault I guess. Maybe it is time to optimize the tests. I think a lot of the tests can be unified to a single big test.

File: compiler/rodutils.nim [hunk 6]
```diff
@@ -22,8 +22,8 @@ proc toStrMaxPrecision*(f: BiggestFloat, literalPostfix = ""): string =
     else: result = "-INF"
   else:
     var buf: array[0..80, char]
-    c_sprintf(buf, "%#.16e" & literalPostfix, f)
-    result = $buf
+    let newLen = c_snprintf(buf.cstring, buf.len.uint, "%#.16e%s", f, literalPostfix.cstring)
```
Comment (Araq) [inline compiler/rodutils.nim]: Use ``discard`` here?

File: lib/system.nim [hunk 7]
```diff
@@ -2410,42 +2410,52 @@ proc `$`*[T: tuple|object](x: T): string =
...
-      if value.isNil: result.add "nil"
-      else: result.add($value)
+      # this branch should not be necessary
+      if value.isNil:
+        result.add "nil"
+      else:
+        result.add($value)
+    # prevent temporary string allocation
+    elif compiles(result.add(value)):
+      # don't insert '\0' characters into the result string
```
Comment (Araq) [inline lib/system.nim]: Huh? Get rid of this special case hack. A ``$`` that skips binary zeros is unusable for anything but ``echo``.

File: lib/system.nim [hunk 8]
```diff
@@ -2410,42 +2410,52 @@ proc `$`*[T: tuple|object](x: T): string =
...
+      when value is char:
```
Comment (Araq) [inline lib/system.nim]: Huh? Get rid of this special case hack. A ``$`` that skips binary zeros is unusable for anything but ``echo``.

File: compiler/rodutils.nim (hunk 6 above)
Comment (krux02) [inline compiler/rodutils.nim]: ok no problem.

File: lib/system.nim (hunk 8 above)
Comment (krux02) [inline lib/system.nim]: Well The reason I put it in there is this:

```
type
  MyType = object
    x: array[10,char]
    a,b,c: int
    s: string
var mt: MyType
echo mt
```
The result string will just end at the first character of the array. I don't think that is very usable, therefore I decided to simply not add null characters to the string. And the string that is created now is reversable. Whenever there are two separators in the result string without a character in between, the character was a null character. So you can think of it like the null character is converted into the empty string. So I think your point that this `$` operator would only be usable for `echo` is not quite true. But if you really don't like this behavior, I can change it like you have. I just think an `$` operator that returns "[a,b,c,,,,,,,]" on an `array[10,char]` is more useful than an operator that returns "[a,b,c,"

Comment (Araq) [inline lib/system.nim]: ``echo`` needs to be fixed to care about the string's length, not it's binary zeros in it.

**Comment by krux02** [comment]
I reverted the ``genEcho`` code to what it was originally. If you want the implementation of ``echo`` to be printed it will not be part of this pull request, because it is not part of what this pull request tries to achieve. ``stdout.writeLine`` can print a string including '\0' characters.

File: lib/system.nim [hunk 9]
```diff
@@ -1852,7 +1852,7 @@ proc `$` *(x: float): string {.magic: "FloatToStr", noSideEffect.}
 proc `$` *(x: bool): string {.magic: "BoolToStr", noSideEffect.}
   ## The stringify operator for a boolean argument. Returns `x`
   ## converted to the string "false" or "true".
-
+#
```
Comment (dom96) [inline lib/system.nim]: Remove this artifact

File: lib/system.nim (hunk 4 above)
Comment (dom96) [inline lib/system.nim]: You mean the ``join`` procedure?

Comment (dom96) [inline lib/system.nim]: I don't think you even need the ``prefix`` or ``suffix``. Just append/prepend in the procs below.

File: tests/async/tnewasyncudp.nim [hunk 10]
```diff
@@ -54,15 +54,15 @@ proc launchSwarm(name: ptr SockAddr) {.async.} =
...
+      zeroMem(cast[pointer](addr(saddr)), sizeof(Sockaddr_in))
       var message = "Message " & $(i * messagesToSend + k)
       await sendTo(sock, addr message[0], len(message),
                    name, sizeof(Sockaddr_in).SockLen)
       var size = await recvFromInto(sock, cast[pointer](addr buffer[0]),
                                     16384, cast[ptr SockAddr](addr saddr),
                                     addr slen)
       size = 0
-      var grammString = $buffer
+      var grammString = $buffer.cstring
```
Comment (dom96) [inline tests/async/tnewasyncudp.nim]: Pretty sure that `buffer` is a `cstring` already... Why is this change here?

Comment (krux02) [inline tests/async/tnewasyncudp.nim]: because buffer is an array and guess what `$` on an array of characters does now.

File: lib/system.nim (hunk 4 above)
Comment (krux02) [reply lib/system.nim]: @dom96 yes in fact I do mean join. And yes ``prefix``/``suffix`` is not required, but still nice to have.
